datastore/sql/sqliteds.nim

```diff
@@ -1,237 +1,191 @@
-          return failure(newRollbackError(rbErr, err))
-        return failure(err)
-
-    var changes = 0.int64
-    proc onChangesResult(s: RawStmtPtr) =
-      changes = changesCol(s, 0)()
-
-    if err =? self.db.getChangesStmt.query((), onChangesResult).errorOption:
-      if rbErr =? self.db.rollbackStmt.exec().errorOption:
-        return failure(newRollbackError(rbErr, err))
-      return failure(err)
+method get*(
+    self: SQLiteDatastore, key: Key
+): Future[?!RawRecord] {.async: (raises: [CancelledError]).} =
+  var
+    rowFound = false
+    value: seq[byte]
+    token = 0'i64
 
-    if changes == 1:
-      if err =? self.db.endStmt.exec().errorOption:
-        if rbErr =? self.db.rollbackStmt.exec().errorOption:
-          return failure(newRollbackError(rbErr, err))
-        return failure(err)
-      break
-    elif changes == 0:
-      if rbErr =? self.db.rollbackStmt.exec().errorOption:
-        return failure(newRollbackError(rbErr, "Unable to retry after race condition was detected"))
-      retriesLeft.dec
-    else:
-      let msg = "Unexpected number of changes, expected either 0 or 1, was " & $changes
-      if rbErr =? self.db.rollbackStmt.exec().errorOption:
-        return failure(newRollbackError(rbErr, msg))
-      return failure(msg)
+  proc onRow(s: RawStmtPtr) =
+    rowFound = true
+    value = dataCol(s, GetSingleStmtDataCol)()
+    token = versionCol(s, GetSingleStmtVersionCol)()
 
-  if retriesLeft == 0:
-    return failure("Retry limit exceeded")
+  if err =? self.db.getSingleStmt.query((key.id), onRow).errorOption:
+    return failure(err)
 
-  return success(aux)
+  if not rowFound:
+    return failure(newException(DatastoreKeyNotFound, "Key doesn't exist"))
 
+  if token < 0:
+    return failure(newCorruptionError("Negative token stored for " & $key))
 
-method modify*(self: SQLiteDatastore, key: Key, fn: Modify): Future[?!void] {.async: (raises: [CancelledError]).} =
-  proc wrappedFn(maybeValue: ?seq[byte]): Future[(?seq[byte], seq[byte])] {.async.} =
-    let res = await fn(maybeValue)
-    let ignoredAux = newSeq[byte]()
-    return (res, ignoredAux)
+  return success RawRecord.init(key, value, token.uint64)
 
-  if err =? (await self.modifyGet(key, wrappedFn)).errorOption:
-    return failure(err)
-  else:
-    return success()
+method get*(
+    self: SQLiteDatastore, keys: seq[Key]
+): Future[?!seq[RawRecord]] {.async: (raises: [CancelledError]).} =
+  var records: seq[RawRecord]
 
-method has*(self: SQLiteDatastore, key: Key): Future[?!bool] {.async: (raises: [CancelledError]).} =
-  var
-    exists = false
+  proc onRow(s: RawStmtPtr) =
+    let
+      key = idCol(s, GetManyStmtIdCol)()
+      value = dataCol(s, GetManyStmtDataCol)()
+      token = versionCol(s, GetManyStmtVersionCol)()
 
-  proc onData(s: RawStmtPtr) =
-    exists = sqlite3_column_int64(s, ContainsStmtExistsCol.cint).bool
+    records.add(RawRecord.init(?Key.init(key), value, token.uint64))
```

Comment:
There's a check for negative tokens in the `get` overload above, but that is missing here.
So if a negative token can really occur, it'll be silently cast to a uint64 here.

Reply:
I think the casting does a check for underflows, but I'll double-check.